src/main/java/io/supertokens/session/accessToken/AccessTokenSigningKey.java

```diff
@@ -69,142 +73,206 @@ public static AccessTokenSigningKey getInstance(Main main) {
+
+        // This reference comparison should work, since we recreate the list object each time we refresh and it's unmodifiable
+        if (this.validKeys == oldKeyInfo) {
             // key has not changed since we previously tried to use it.. So we can make it null.
             // otherwise we might end up making this null unnecessarily.
 
             ProcessState.getInstance(this.main)
                     .addState(ProcessState.PROCESS_STATE.SETTING_ACCESS_TOKEN_SIGNING_KEY_TO_NULL, null);
-            this.keyInfo = null;
+            this.validKeys = null;
         }
     }
 
-    public synchronized AccessTokenSigningKey.KeyInfo getKey()
+    public synchronized List<KeyInfo> getKey()
             throws StorageQueryException, StorageTransactionLogicException {
-        if (this.keyInfo == null) {
-            this.keyInfo = maybeGenerateNewKeyAndUpdateInDb();
+        if (this.validKeys == null || this.validKeys.size() == 0) {
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
         }
 
         CoreConfig config = Config.getConfig(main);
-        if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > (this.keyInfo.createdAtTime
+        if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > (this.validKeys.get(0).createdAtTime
                 + config.getAccessTokenSigningKeyUpdateInterval())) {
             // key has expired, we need to change it.
-            this.keyInfo = maybeGenerateNewKeyAndUpdateInDb();
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
         }
-        return this.keyInfo;
+        return this.validKeys;
+    }
+
+    public KeyInfo getCurrentKey()
+            throws StorageQueryException, StorageTransactionLogicException {
+        return this.getKey().get(0);
     }
 
     public synchronized long getKeyExpiryTime() throws StorageQueryException, StorageTransactionLogicException {
         if (Config.getConfig(this.main).getAccessTokenSigningKeyDynamic()) {
             this.getKey();
-            long createdAtTime = this.keyInfo.createdAtTime;
+            // getKey ensures we have at least 1 valid keys
+            long createdAtTime = this.validKeys.get(0).createdAtTime;
             return createdAtTime + Config.getConfig(main).getAccessTokenSigningKeyUpdateInterval();
         } else {
             // return 10 years from now
             return System.currentTimeMillis() + (10L * 365 * 24 * 3600 * 1000);
         }
     }
 
-    private KeyInfo maybeGenerateNewKeyAndUpdateInDb() throws StorageQueryException, StorageTransactionLogicException {
+    private List<KeyInfo> maybeGenerateNewKeyAndUpdateInDb() throws StorageQueryException, StorageTransactionLogicException {
         Storage storage = StorageLayer.getSessionStorage(main);
         CoreConfig config = Config.getConfig(main);
+        
+        final long validityDuration = config.getAccessTokenSigningKeyUpdateInterval() + 2 * config.getAccessTokenValidity();
```

Comment:
- Store the "2" in a constant somewhere on top of this class (as a `final static` variable).
- Also, should we make this configurable by the user? Cause if someone has an app where most people access it at a certain time of the day, and that time is such that it is greater than `validityDuration + timeLatesKeyWasCreated`, there would still be a spike in the refresh APIs.